### PR TITLE
Use argparse for arg parsing

### PR DESCRIPTION
**Pull request by chriselion**
Replace docopt with python's argparsing library.

``` console
$ mlagents-learn --help

...

usage: mlagents-learn [-h] [--env ENV_PATH] [--curriculum CURRICULUM_FOLDER]
                      [--sampler SAMPLER_FILE_PATH]
                      [--keep-checkpoints KEEP_CHECKPOINTS] [--lesson LESSON]
                      [--load] [--run-id RUN_ID] [--num-runs NUM_RUNS]
                      [--save-freq SAVE_FREQ] [--seed SEED] [--slow] [--train]
                      [--base-port BASE_PORT] [--num-envs NUM_ENVS]
                      [--docker-target-name DOCKER_TARGET_NAME]
                      [--no-graphics] [--debug] [--multi-gpu]
                      trainer_config_path

positional arguments:
  trainer_config_path

optional arguments:
  -h, --help            show this help message and exit
  --env ENV_PATH        Name of the Unity executable (default: None)
  --curriculum CURRICULUM_FOLDER
                        Curriculum json directory for environment (default:
                        None)
  --sampler SAMPLER_FILE_PATH
                        Reset parameter yaml file for environment (default:
                        None)
  --keep-checkpoints KEEP_CHECKPOINTS
                        How many model checkpoints to keep (default: 5)
  --lesson LESSON       Start learning from this lesson (default: 0)
  --load                Whether to load the model or randomly initialize
                        (default: False)
  --run-id RUN_ID       The directory name for model and summary statistics
                        (default: ppo)
  --num-runs NUM_RUNS   Number of concurrent training sessions (default: 1)
  --save-freq SAVE_FREQ
                        Frequency at which to save model (default: 50000)
  --seed SEED           Random seed used for training (default: -1)
  --slow                Whether to run the game at training speed (default:
                        False)
  --train               Whether to train model, or only run inference
                        (default: False)
  --base-port BASE_PORT
                        Base port for environment communication (default:
                        5005)
  --num-envs NUM_ENVS   Number of parallel environments to use for training
                        (default: 1)
  --docker-target-name DOCKER_TARGET_NAME
                        Docker volume to store training-specific files
                        (default: None)
  --no-graphics         Whether to run the environment in no-graphics mode
                        (default: False)
  --debug               Whether to run ML-Agents in debug mode with detailed
                        logging (default: False)
  --multi-gpu           Setting this flag enables the use of multiple GPU's
                        (if available) during training (default: False)
```

### CODE REVIEW

**Comment by chriselion** [comment]
This looks like a big change but it was pretty well tested along the way. The basic process was
* define the CommandLineOptions namedtuple and from that from the docopt code
* add unit tests that the CommandLineOptions parse as expected
* write the argparse implementation and get it passing the same tests
* remove docopt implementation and clean up

**Comment by chriselion** [comment]
This is a prerequisite for doing https://github.com/Unity-Technologies/ml-agents/pull/2491 the "right" way, but doesn't accomplish that here.

File: ml-agents/mlagents/trainers/learn.py [hunk 1]
```diff
@@ -23,104 +23,218 @@
...
+    lesson: int
+    slow: bool
+    no_graphics: bool
+    multi_gpu: bool  # ?
+    trainer_config_path: str
+    sampler_file_path: Optional[str]
+    docker_target_name: Optional[str]
+
+    @property
+    def fast_simulation(self) -> bool:
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/learn.py]: Note that this is controlled with `--slow`, but seemed easiest to do this way.

File: ml-agents/mlagents/trainers/learn.py [hunk 2]
```diff
@@ -23,104 +23,218 @@
...
-        run_options["--sampler"] if run_options["--sampler"] != "None" else None
-    )
+    trainer_config_path = options.trainer_config_path
+    curriculum_folder = options.curriculum_folder
 
     # Recognize and use docker volume if one is passed as an argument
-    if not docker_target_name:
-        model_path = "./models/{run_id}-{sub_id}".format(run_id=run_id, sub_id=sub_id)
+    if not options.docker_target_name:
+        model_path = "./models/{run_id}-{sub_id}".format(
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/learn.py]: Could consider making these properties of the `CommandLineOptions` class but didn't want to change too much here at once.

**Comment by surfnerd** [comment]
I'm not a reviewer on this but 👍 for removing a pip dep and yay for argparse

File: ml-agents/mlagents/trainers/tests/test_learn.py [hunk 3]
```diff
@@ -84,3 +70,73 @@ def test_docker_target_path(
             mock_init.assert_called_once()
             assert mock_init.call_args[0][1] == "/dockertarget/models/ppo-0"
             assert mock_init.call_args[0][2] == "/dockertarget/summaries"
+
+
+def test_commandline_args():
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/tests/test_learn.py]: 👍 for tests

File: docs/Training-on-Microsoft-Azure-Custom-Instance.md [hunk 4]
```diff
@@ -85,7 +85,6 @@ This page contains instructions for setting up a custom Virtual Machine on Micro
     ```sh
     pip3 install pillow
     pip3 install numpy
-    pip3 install docopt
```
Comment (awjuliani) [inline docs/Training-on-Microsoft-Azure-Custom-Instance.md]: 👍

File: ml-agents/setup.py [hunk 5]
```diff
@@ -31,7 +31,6 @@
     zip_safe=False,
     install_requires=[
         # Test-only dependencies should go in test_requirements.txt, not here.
-        "docopt",
```
Comment (awjuliani) [inline ml-agents/setup.py]: 👍

**Pull request merged**